docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,112 @@
+* I want to remove OpenShift `Groups` that were previously backed by LDAP records that no longer exist
+* I want to specify an explicit mapping from LDAP group identifiers to OpenShift `Group` resources
+
+## Supported LDAP Group Schemas
+The three LDAP schemas supported are
+
+1. groups as first-class LDAP entries that contain lists of members of the group
+2. users as first-class LDAP entries that contain lists of groups they are a member of
+3. schema 2 above, also including first-class groups to store metadata
+
+### Required OpenShift `Group` Metadata
+OpenShift `Group`'s `Annotations` will be used to store metadata regarding the sync process. `Annotations` will store the LDAP unique identifier (`openshift.io/ldap-uid`), the last sync time (`openshift.io/ldap-last-sync-time`), and, if available, the LDAP resource version (`openshift.io/ldap-resource-version`). A `Label` will also be set to determine the external record provider (`openshift.io/external-record-provider=LDAP`).
+
+### Data Flow
+* Determine ordered list of LDAP groups to be synced
+* Find LDAP group members
+* Determine OpenShift `Identities` from LDAP group members
+* Determine OpenShift `Users` from OpenShift `Identities`
+* Determine OpenShift `Group` from LDAP group
+* Populate new OpenShift `Group`'s `Users` from resulting OpenShift `UserIdentityMappings`, update OpenShift `Group` metadata fields tied to LDAP attributes, leave other OpenShift `Group` fields unchanged
+
+##  Determining Which LDAP Groups To Sync
+An `LDAPGroupLister` determines what LDAP groups needs to be synced and outputs the result as a set of unique identifier strings (called "LDAP group UIDs" in this document). The other objects that take LDAP group UIDs need to understand the format of this string (e.g. these objects will be tightly coupled). For example, an `LDAPGroupLister` for schema 1 above cannot be used with a `LDAPGroupDataExtractor` for schema 2. The four `LDAPGroupLister` implementations that are necessary are:
+
+1. List all OpenShift `Groups`, minus a blacklist
+* List some whitelist of OpenShift `Groups`
+* List all LDAP groups, minus a blacklist
+* List some whitelist of LDAP groups
+
+```go
+// LDAPGroupLister lists the LDAP groups that need to be synced by a job. The LDAPGroupLister needs to
+// be paired with an LDAPGroupDataExtractor that understands the format of the unique identifiers returned
+// to represent the LDAP groups to be synced.
+type LDAPGroupLister interface {
+	ListGroups() (groupUIDs []string, err error)
+}
+```
+
+## Collecting LDAP Members and Metadata
+An `LDAPGroupDataExtractor` gathers information on an LDAP group based on an LDAP group UID. It may cache LDAP responses for responsivity. The approach to implementing this structure will vary with LDAP schema as well as sync job request. There will be a specific implementation for each schema as well as another flavor of that implementation for that schema where all the LDAP groups need to be synced.
```

Comment:
Kill this bit: "as well as another flavor of that implementation for that schema where all the LDAP groups need to be synced"